peps/pep-806.rst

```diff
@@ -0,0 +1,292 @@
+2. **Multiple context managers**: Show the current comma syntax
+3. **Async context managers**: Introduce ``async with``
+4. **Mixed contexts**: "Mark each async context manager with ``async``"
+
+
+Rejected Ideas
+==============
+
+Workaround: an ``as_acm()`` wrapper
+-----------------------------------
+
+It is easy to implement a helper function which wraps a synchronous context
+manager in an async context manager.  For example:
+
+.. code-block:: python
+
+    @contextmanager
+    async def as_acm(sync_cm):
+        with sync_cm as result:
+            await sleep(0)
+            yield result
+
+    async with (
+        acquire_lock(),
+        as_acm(open('file')) as f,
+    ):
+        ...
+
+This is our recommended workaround for almost all code.
+
+However, there are some cases where calling back into the async runtime (i.e.
+executing ``await sleep(0)``) to allow cancellation is undesirable.  On the
+other hand, *omitting* ``await sleep(0)`` would break the transitive property
+that a syntactic ``await`` / ``async for`` / ``async with`` always calls back
+into the async runtime (or raises an exception).  While few codebases enforce
+this property, we have found it indispensible in preventing deadlocks.
+
+
+Workaround: using ``AsyncExitStack``
+------------------------------------
+
+:class:`~contextlib.AsyncExitStack` offers a powerful, low-level interface
+which allows for explicit entry of sync and/or async context managers.
+
+.. code-block:: python
+
+    async with contextlib.AsyncExitStack() as stack:
+        await stack.enter_async_context(acquire_lock())
+        f = stack.enter_context(open('file'))
+        ...
+
+However, :class:`~contextlib.AsyncExitStack` introduces significant complexity
+and potential for errors - it's easy to violate properties that syntactic use
+of context managers would guarantee, such as 'last-in, first-out' order.
+
+
+Workaround: ``AsyncExitStack``-based helper
+-------------------------------------------
+
+We could also implement a ``multicontext()`` wrapper, which avoids some of the
+downsides of direct use of :class:`~contextlib.AsyncExitStack`:
+
+.. code-block:: python
+
+    async with multicontext(
+        acquire_lock(),
+        open('file'),
+    ) as (f, _):
+        ...
+
+However, this helper breaks the locality of ``as`` clauses, which makes it
+easy to accidentally mis-assign the yielded variables (as in the code sample).
+It also requires either distinguishing sync from async context managers using
+something like a tagged union - perhaps overload an operator so that, e.g.,
+``async_ @ acquire_lock()`` works - or else guessing what to do with objects
+that implement both sync and async context-manager protocols.
+Finally, it has the error-prone semantics around exception handling which led
+`contextlib.nested()`__ to be deprecated in favor of the multi-argument
+``with`` statement.
+
+__ https://docs.python.org/2.7/library/contextlib.html#contextlib.nested
```

Comment:
We could add intersphinx for 2.7 here?

Reply:
I think (hope?) that references to Python 2.7 documentation in new PEPs are rare enough that adding intersphinx is not worth the overhead.